### PR TITLE
Add a `Receiver.triggered` method

### PR DESCRIPTION
**Pull request by shsms**
This is an alternative to the `selected_from` method.

### CODE REVIEW

**Comment by shsms** [comment]
Other naming options: `matches`, `matches_selected`

File: src/frequenz/channels/_receiver.py [hunk 1]
```diff
@@ -155,11 +155,14 @@
...
 from abc import ABC, abstractmethod
 from collections.abc import Callable
-from typing import Generic, Self
+from typing import TYPE_CHECKING, Any, Generic, Self, TypeGuard
 
 from ._exceptions import Error
 from ._generic import MappedMessageT_co, ReceiverMessageT_co
 
+if TYPE_CHECKING:
+    from ._select import Selected
```
Comment (llucax) [inline src/frequenz/channels/_receiver.py:164]: :unamused: 

Oh, this is also an ugly side-effect of moving part of the implementation of `select` to `Receiver`. We could consider moving `Selected` to this file, if we are mixing stuff up anyway.

**Review by llucax — COMMENTED** [review]
OK, so the issue we had with `TypeGuard` as a member method was not it being a member method but about [applying the type guard to `self`](https://github.com/microsoft/pyright/discussions/3125), this is why we can't do `selected.source(receiver)`, because we want to narrow `selected`, not `receiver`, but when used as a member the variable to narrow is the second argument, not `self`.

So `receiver.is_selected(selected)` works and will narrow `selected. OK. I don't like the fact that now part of `select()` is implemented in `Receiver` unless we have a considerable improvement in how the code is read, and `receiver.is_selected(selected)` doesn't read much better than `selected_from(selected, receiver)`.

But maybe something like `receiver.is_source_of(selected)` could work. What do you think?

File: src/frequenz/channels/_receiver.py (hunk 1 above)
Comment (shsms) [reply src/frequenz/channels/_receiver.py:164]: True, maybe also rename it so it sounds like `WrappedMessage` or `ReceivedMessage`, and it would be decoupled from `select` completely?

If not, it feels a bit weird, defining a `select` feature in the receiver module.

Comment (llucax) [reply src/frequenz/channels/_receiver.py:164]: I don't know, for me it is unfortunate that we need to implement part of `select()` in receiver but it doesn't make sense to use weirder names only to try to hide that. `Selected` / `WrappedMessage` / `ReceivedMessage` will only be used by select anyways, I would keep calling it `Selected`.

I really have double feelings about this PR, for one side I think it improves the UX slightly, but I'm not sure if enough to worth a deprecation and pushing people to update their code, and for the other side I think it makes the implementation slightly less clean (more coupling between `Receiver` and `select`), but I'm not sure if enough to completely oppose to the change :smile: :disappointed:

Comment (shsms) [reply src/frequenz/channels/_receiver.py:164]: > I think it makes the implementation slightly less clean (more coupling between `Receiver` and `select`)

I would look at it as an increased commitment to having `select` as a core component of the channels package, select is very deeply related to receivers.

And if it was a different language, we wouldn't even notice the coupling, because no language makes so much noise about these things as python.

Comment (shsms) [reply src/frequenz/channels/_receiver.py:164]: > I think it improves the UX slightly,

I think it is more than a slight improvement for new users, because they don't have to import the weird-looking `selected_from` separately, so there's less confusion.

Comment (Marenz) [reply src/frequenz/channels/_receiver.py:164]: I think we can support the the old way for long to stay backwards compatible to avoid annoying users though?

Comment (shsms) [reply src/frequenz/channels/_receiver.py:164]: Oh right, so we can't rename `Selected`.

Also you mean add the new stuff, but don't deprecate the `selected_from` method?

Comment (llucax) [reply src/frequenz/channels/_receiver.py:164]: Yeah, maybe we can deprecate it after a few more releases, when the new form is more widely used, to confirm it is really better than the old form.

Comment (llucax) [reply src/frequenz/channels/_receiver.py:164]: > Oh right, so we can't rename `Selected`.

We can always rename it and create a (deprecated) alias using the old name. But personally I wouldn't do it, I think that name is fine.

Anyway, as I said, my main complain about this PR is the current `is_selected()` name, if we get a better name I'm OK with merging. I think my preferred name is `is_source_of()` but I can also live with `matches()`

Comment (llucax) [reply src/frequenz/channels/_receiver.py:164]: About names, I must admit I also thought about `was`, which sucks as a name taken out of context, but reads very nicely when put in `receiver.was(selected)` :smile: :grimacing: 
ChatGPT also suggested `receiver.is_active(selected)` which doesn't reads as nicely as `receiver.is_actively(selected)` but makes much more sense as a function name.

EDIT: One issue with `is_active()` is it sounds much less obvious that you should always call it, even if you then ignore the receiver, but I guess that's true for most options, but I think it is better captured by `selected_from` or  `receiver.was(selected)`. I guess `matches` is also better than `is_active` or `is_source_of` in this sense. Since this is a very weird construct anyways, I'm really considering `receiver.was(selected)` as an option...

Comment (shsms) [reply src/frequenz/channels/_receiver.py:164]: Similar to `was`, but maybe a bit more obvious, and independent of the variable naming conventions for `selected`, maybe `receiver.produced(selected)` is an option.  `was` doesn't work IMO, because it depends on the variable name `selected`, which we don't have any control over.

But I'm still in favour of `matches`, because it has the coolest sound, with a lispy/functional feel, in a select block:

```python
async for selected in select(power_rx, timer):
    if power_rx.matches(selected):
        print(f"Power: {power}")
    elif timer.matches(selected):
        print("Timer triggered")
```

Comment (llucax) [reply src/frequenz/channels/_receiver.py:164]: In the line of `produced`, what about `triggered`?

```py
async for selected in select(power_rx, timer):
    if power_rx.triggered(selected):
        print(f"Power: {power}")
    elif timer.triggered(selected):
        print("Timer triggered")
```

What it bothers me about `matches` is it is really the `selected` that matches the receiver, not the other way around (if we could use `selected.matches(receiver)` that would be a :100: ). But as I said, I can live with `matches`.

Comment (shsms) [reply src/frequenz/channels/_receiver.py:164]: You were right.  I'm also having doubts about `matches`.  I will update the PR to say `triggered`.

Comment (llucax) [reply src/frequenz/channels/_receiver.py:164]: :scream:

File: RELEASE_NOTES.md [hunk 2]
```diff
@@ -10,7 +10,15 @@
...
 
-<!-- Here goes the main new features and examples or instructions on how to use them -->
+- There is a new `Receiver.triggered` method that can be used instead of `selected_from`:
+
+  ```python
+  async for selected in select(recv1, recv2):
+      if recv1.triggered(selected):
+          print('Received from recv1:', selected.value)
+      if recv2.triggered(selected):
+          print('Received from recv2:', selected.value)
```
Comment (llucax) [inline RELEASE_NOTES.md]:
```suggestion
          print('Received from recv1:', selected.message)
      if recv2.triggered(selected):
          print('Received from recv2:', selected.message)
```

File: src/frequenz/channels/_receiver.py [hunk 3]
```diff
@@ -373,9 +400,7 @@ def consume(self) -> MappedMessageT_co:  # noqa: DOC502
             ReceiverStoppedError: If the receiver stopped producing messages.
             ReceiverError: If there is a problem with the receiver.
         """
-        return self._mapping_function(
-            self._receiver.consume()
-        )  # pylint: disable=protected-access
+        return self._mapping_function(self._receiver.consume())
```
Comment (llucax) [inline src/frequenz/channels/_receiver.py:403]: Optional: This is unrelated to the commit, could be moved to another commit.

File: src/frequenz/channels/_select.py [hunk 4]
```diff
@@ -378,7 +376,7 @@ async def select(  # noqa: DOC503
         import datetime
         from typing import assert_never
 
-        from frequenz.channels import ReceiverStoppedError, select, selected_from
```
Comment (llucax) [inline src/frequenz/channels/_select.py:381]: This shouldn't be removed, or the example needs to be updated to use `Receiver.triggered()` instead.

File: RELEASE_NOTES.md (hunk 2 above)
Comment (Marenz) [reply RELEASE_NOTES.md]: Shouldn't sybil catch this? Ah, probably because it's a RELEASE_NOTES.md file.

Comment (llucax) [reply RELEASE_NOTES.md]: We are not "testing" the `RELEASE_NOTES.md`. I think we discussed it at some point, my feeling is no, the release notes sometimes have also old code as example of how to migrate some code, that it is not intended to work. I think it would be difficult to test examples in the release notes.

Comment (shsms) [reply RELEASE_NOTES.md]: fixed.

File: src/frequenz/channels/_receiver.py (hunk 3 above)
Comment (shsms) [reply src/frequenz/channels/_receiver.py:403]: done.

File: src/frequenz/channels/_select.py (hunk 4 above)
Comment (shsms) [reply src/frequenz/channels/_select.py:381]: oops, replaced with `Receiver.triggered()`

**Review by llucax — APPROVED** [review]
